File: document-processor/app/services/sqs_service.py
import asyncio
import json
from typing import Awaitable, Callable, Dict, Any, Optional
import boto3
from app.core.config import settings
from app.services.document_processor import DocumentProcessor
from app.core import util

logger = util.init_logger(__name__)


class SQSService:
    """Service for handling SQS queue operations"""

    def __init__(self):
        self.sqs_client = None
        self.queue_url = settings.sqs_documents_queue
        self.index_queue_url = settings.sqs_index_documents_queue
        self.running = False
        self.poll_task = None
        self.processor: Optional[DocumentProcessor] = None

    async def startup(self):
        """Initialize SQS client and start polling"""
        try:
            self.sqs_client = boto3.client(
                "sqs",
                endpoint_url=settings.sqs_endpoint or None,
                region_name=settings.sqs_region,
                aws_access_key_id=settings.sqs_access_key_id or None,
                aws_secret_access_key=settings.sqs_secret_access_key or None,
            )
            self.running = True
            self.processor = DocumentProcessor(self.send_message)
            await self.processor.startup()

            # TDOD: Run settings.num_threads pollers with MaxNumberOfMessages=1 
            # each for better parallelism
            self.poll_task = asyncio.create_task(self._poll_messages())

            logger.info(f"SQS Service started, polling queue: {self.queue_url}")
        except Exception as e:
            logger.error(f"Failed to start SQS service: {e}")
            raise

    async def shutdown(self):
        """Stop polling and cleanup resources"""
        logger.info("SQS Service shutting down...")
        self.running = False
        if self.poll_task:
            self.poll_task.cancel()
            try:
                await self.poll_task
            except asyncio.CancelledError:
                logger.debug("Poll task cancelled successfully")
            self.poll_task = None
        if self.processor:
            await self.processor.shutdown()
        if self.sqs_client:
            self.sqs_client.close()
        logger.info("SQS Service stopped")

    async def _poll_messages(self):
        """Poll for messages from SQS queue"""
        response = self.sqs_client.list_queues(MaxResults=100)
        logger.debug(f"Found queues: {response}")

        while self.running:
            try:
                # Use asyncio.to_thread to make the blocking call cancellable
                # and run it in a thread pool so it doesn't block the event loop
                response = await asyncio.to_thread(
                    self.sqs_client.receive_message,
                    QueueUrl=self.queue_url,
                    MaxNumberOfMessages=settings.num_threads,
                    WaitTimeSeconds=5,  # Shorter polling for better shutdown responsiveness
                    VisibilityTimeout=120,  # seconds to process
                )
                if not self.running:
                    logger.debug("Polling stop requested; exiting loop")
                    break
                messages = response.get("Messages", [])
                # Process all messages in parallel
                if messages:
                    logger.debug(f"Got messages: {messages}")

                    async def process_message(message):
                        try:
                            sqs_client = self.sqs_client
                            queue_url = self.queue_url
                            def ack():
                                # Delete message from queue after processing
                                sqs_client.delete_message(
                                    QueueUrl=queue_url,
                                    ReceiptHandle=message["ReceiptHandle"],
                                )
                                logger.info(f"Deleted message from SQS: {message['MessageId']}")
                            
                            # Process message
                            await self._handle_message(message, ack=ack)

                        except Exception as e:
                            logger.error(f"Error processing message: {message}")
                            logger.exception(e, exc_info=True)
                            # Message will become visible again after timeout

                    await asyncio.gather(
                        *[process_message(msg) for msg in messages],
                        return_exceptions=True,
                    )

            except asyncio.CancelledError:
                logger.info("Polling cancelled, shutting down")
                break
            except Exception as e:
                logger.error(f"Error polling SQS: {e}")
                await asyncio.sleep(5)  # Wait before retrying

    async def _handle_message(self, message: Dict[str, Any], ack: Callable[[], None]):
        """Handle incoming SQS message"""
        try:
            body = json.loads(message["Body"])
            logger.info(f"Processing message: {body}")
            await self.processor.handle_command(body, ack=ack)

        except json.JSONDecodeError as e:
            logger.error(f"Invalid JSON in message body: {e}")
            raise

    async def send_message(
        self, is_processing: bool, command: Dict[str, Any], delay_seconds: int = 0
    ):
        """Send a message to the SQS queue"""
        try:
            await asyncio.to_thread(
                self.sqs_client.send_message,
                QueueUrl=self.queue_url if is_processing else self.index_queue_url,
                MessageBody=json.dumps(command),
                DelaySeconds=delay_seconds,
            )
            logger.info(f"Sent message to SQS: {command}")
        except Exception as e:
            logger.error(f"Failed to send message to SQS: {e}")
            raise

File: api/src/services/sqs.service.ts
import { createLogger } from "@/utils/logger";
import { DocumentQueueService } from "./document-queue.service";
import {
  SQS_ACCESS_KEY_ID,
  SQS_DOCUMENTS_QUEUE,
  SQS_INDEX_DOCUMENTS_QUEUE,
  SQS_ENDPOINT,
  SQS_REGION,
  SQS_SECRET_ACCESS_KEY,
} from "@/config/application";
import {
  SQSClient,
  ReceiveMessageCommand,
  DeleteMessageCommand,
  SendMessageCommand,
  Message,
  ListQueuesCommand,
} from "@aws-sdk/client-sqs";
import { SubscriptionsService } from "./subscriptions.service";

const logger = createLogger(__filename);

export class SQSService {
  private sqs: SQSClient;
  private outputQueueUrl: string;
  private indexQueueUrl: string;
  private documentQueueService: DocumentQueueService;
  private polling = false;
  private pollInterval?: NodeJS.Timeout;

  constructor(subscriptionsService: SubscriptionsService) {
    this.outputQueueUrl = SQS_DOCUMENTS_QUEUE || "";
    this.indexQueueUrl = SQS_INDEX_DOCUMENTS_QUEUE || "";
    this.documentQueueService = new DocumentQueueService(subscriptionsService);

    // Initialize SQS client
    logger.info({ SQS_ENDPOINT, SQS_ACCESS_KEY_ID, SQS_REGION }, "Initializing SQS client");
    this.sqs = new SQSClient({
      endpoint: SQS_ENDPOINT || undefined,
      region: SQS_REGION || "us-east-1",
      credentials:
        SQS_ACCESS_KEY_ID && SQS_SECRET_ACCESS_KEY
          ? {
              accessKeyId: SQS_ACCESS_KEY_ID,
              secretAccessKey: SQS_SECRET_ACCESS_KEY,
            }
          : undefined,
    });
  }

  async startup(): Promise<void> {
    if (!this.indexQueueUrl) {
      logger.warn("SQS_INDEX_DOCUMENTS_QUEUE not configured, skipping SQS listener");
      return;
    }

    this.polling = true;
    this.startPolling();

    logger.info(`API SQS Service started, polling queue: ${this.indexQueueUrl}`);
  }

  async shutdown(): Promise<void> {
    if (!this.polling) return;
    this.sqs.destroy();
    this.polling = false;
    if (this.pollInterval) {
      clearTimeout(this.pollInterval);
    }
    logger.info("API SQS Service stopped");
  }

  /**
   * Send an arbitrary JSON string message to the SQS queue
   */
  async sendMessage(
    messageBody: string,
    delaySeconds: number = 0,
    indexQueue: boolean = false
  ): Promise<string | undefined> {
    const queueUrl = indexQueue ? this.indexQueueUrl : this.outputQueueUrl;

    if (!queueUrl) {
      logger.warn(`SQS queue ${queueUrl} not configured, message not sent`);
      return;
    }
    if (!this.sqs) {
      throw new Error("SQS client not initialized.");
    }

    try {
      const command = new SendMessageCommand({
        QueueUrl: queueUrl,
        MessageBody: messageBody,
        DelaySeconds: delaySeconds || undefined,
      });

      const result = await this.sqs.send(command);

      logger.info(
        {
          messageId: result.MessageId,
          md5: result.MD5OfMessageBody,
          messageLength: messageBody.length,
          delaySeconds,
        },
        "Sent SQS message"
      );

      return result.MessageId;
    } catch (error) {
      logger.error(error, "Failed to send SQS message");
      throw error;
    }
  }

  /**
   * Send a JSON object as a message to the SQS queue
   */
  async sendJsonMessage(
    message: any,
    delaySeconds: number = 0,
    indexQueue: boolean = false
  ): Promise<string | undefined> {
    const messageBody = JSON.stringify(message);
    return this.sendMessage(messageBody, delaySeconds, indexQueue);
  }

  /**
   * Send a delayed message (useful for retry logic)
   */
  async sendDelayedMessage(messageBody: string, delayMinutes: number): Promise<string | undefined> {
    const delaySeconds = Math.min(delayMinutes * 60, 900); // AWS SQS max delay is 15 minutes
    return this.sendMessage(messageBody, delaySeconds);
  }

  private startPolling(): void {
    if (!this.sqs || !this.polling) return;

    const poll = async () => {
      if (!this.polling) {
        return;
      }

      const cmd = new ListQueuesCommand({
        MaxResults: 100,
      });

      const response = await this.sqs.send(cmd);
      if (!response.QueueUrls?.includes(this.indexQueueUrl)) {
        logger.info(`SQS queue ${this.indexQueueUrl} does not exist or is not accessible`);
        clearTimeout(this.pollInterval);
        this.pollInterval = setTimeout(poll, 3000);
        return;
      }
      if (!response.QueueUrls?.includes(this.outputQueueUrl)) {
        logger.info(`SQS processing queue ${this.outputQueueUrl} does not exist or is not accessible`);
        clearTimeout(this.pollInterval);
        this.pollInterval = setTimeout(poll, 3000);
        return;
      }

      try {
        const command = new ReceiveMessageCommand({
          QueueUrl: this.indexQueueUrl,
          MaxNumberOfMessages: 1,
          WaitTimeSeconds: 15,
          VisibilityTimeout: 90,
          AttributeNames: ["All"],
        });

        const result = await this.sqs!.send(command);
        const messages = result.Messages || [];

        logger.debug(`Got ${messages.length} messages from SQS`);

        for (const message of messages) {
          try {
            const handled = await this.handleMessage(message);

            // Delete message after successful processing
            if (handled) {
              await this.deleteMessage(message);
            }
          } catch (error) {
            logger.error(error, "Error processing message");
            // Message will become visible again after timeout
          }
        }
      } catch (error) {
        logger.error(error, `Error polling SQS on ${this.indexQueueUrl}`);
      }

      // Schedule next poll
      if (this.polling) {
        clearTimeout(this.pollInterval);
        this.pollInterval = setTimeout(poll, 100);
      }
    };

    // Start polling
    poll();
  }

  private async handleMessage(message: Message): Promise<boolean> {
    if (!message.Body) {
      logger.warn("Received message without body");
      return true;
    }

    try {
      const command = JSON.parse(message.Body);
      logger.info(`Processing SQS message: ${command.command} for document ${command.documentId}`);

      if (command.command === "index_document") {
        this.documentQueueService
          .handleIndexDocumentCommand(command)
          .then(() => this.deleteMessage(message))
          .catch(error => {
            logger.error(error, `Failed to process index_document command for document ${command.documentId}`);
          });
        return false;
      }

      logger.info(`Skip command: ${command.command}`);
      return false;
    } catch (error) {
      logger.error(error, "Error handling SQS message");
      throw error;
    }
  }

  private async deleteMessage(message: Message) {
    const deleteCommand = new DeleteMessageCommand({
      QueueUrl: this.indexQueueUrl,
      ReceiptHandle: message.ReceiptHandle!,
    });
    await this.sqs.send(deleteCommand);
  }
}
